primitives/message-lane/src/source_chain.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Primitives of message lane module, that are used on the source chain.
+
+use crate::{LaneId, MessageNonce};
+
+use frame_support::Parameter;
+use sp_std::fmt::Debug;
+
+/// Target chain API. Used by source chain, to verify target chain proofs.
+///
+/// All implementations of this trait should only work with finalized data that
+/// can't change. Wrong implementation may lead to invalid lane states (i.e. lane
+/// that's stuck) and/or processing messages without paying fees.
+pub trait TargetHeaderChain<Payload> {
```

Comment:
```suggestion
pub trait TargetHeaderChain<P> {
```

I'd avoid using generics with full names unless it was a particularly ambiguous situation like `pub trait Foo<Hash, Header>` where `H` wouldn't quite cut it.

Reply:
I disagree. Reading a bunch of one-letter generics is a nightmare, it's like using one-letter variables. I'd flip the rule and say we should only use one-letter generic names if it's clear from the context what they mean. In my past work we used to call generics with `T` prefix, to distinguish from regular types, so it would be `TPayload, THash, TFee`, etc. I'd leave the decision to go with one-letter or full-length up to Slava though.

Reply:
So in this case I think it's fine to use a one letter generic (as is Rust convention) since it's only used as a type for an argument called `payload`. When you see `fn foo(payload: P)` it's clear what the generic is for.

I agree that this isn't a blanket rule, and we should be thoughtful when naming our generics. But we should start by following Rust conventions.